python/ray/data/_internal/execution/streaming_executor_state.py

```diff
@@ -181,21 +189,55 @@ def select_operator_to_run(topology: Topology) -> Optional[PhysicalOperator]:
     operators with a large number of running tasks `num_active_tasks()`.
     """
 
-    # TODO: set limits properly based on resources and execution options. This is just
-    # a hard-coded development placeholder.
-    PARALLELISM_LIMIT = 8
-    num_active_tasks = sum(
-        op_state.num_active_tasks() for op_state in topology.values()
-    )
-    if num_active_tasks >= PARALLELISM_LIMIT:
-        return None
+    # Filter to ops that are eligible for execution.
+    ops = [
+        op
+        for op, state in topology.items()
+        if state.num_queued() > 0 and _execution_allowed(op, cur_usage, limits)
+    ]
+
+    # To ensure liveness, allow at least 1 op to run regardless of limits.
+    if not ops and all(op.num_active_work_refs() == 0 for op in topology):
+        # The topology is entirely idle, so choose from all ready ops ignoring limits.
+        ops = [op for op, state in topology.items() if state.num_queued() > 0]
 
-    # Filter to ops that have queued inputs.
-    ops = [op for op, state in topology.items() if state.num_queued() > 0]
+    # Nothing to run.
     if not ops:
         return None
 
     # Equally penalize outqueue length and active tasks for backpressure.
     return min(
         ops, key=lambda op: len(topology[op].outqueue) + topology[op].num_active_tasks()
     )
+
+
+def _execution_allowed(
+    op: PhysicalOperator,
+    global_usage: ExecutionResources,
+    global_limits: ExecutionResources,
+) -> bool:
+    """Return whether an operator is allowed to execute given resource usage.
+
+    Args:
+        op: The operator to check.
+        global_usage: Resource usage across the entire topology.
+        global_limits: Execution resource limits.
+
+    Returns:
+        Whether the op is allowed to run.
+    """
+    # To avoid starvation problems when dealing with fractional resource types,
+    # convert all quantities to integer (0 or 1) for deciding admissibility. This
+    # allows operators with non-integral requests to slightly overshoot the limit.
+    global_floored = ExecutionResources(
+        cpu=math.floor(global_usage.cpu or 0),
+        gpu=math.floor(global_usage.gpu or 0),
+        object_store_memory=global_usage.object_store_memory,
+    )
+    inc = op.incremental_resource_usage()
+    inc_indicator = ExecutionResources(
```

Comment:
It looks we are comparing the resource request (of one additional task) against the global/aggregated resource availability, so passing the check is only a necessary condition (the scheduling is actually not feasible as it will need to find a single node to satisfy the request).

Reply:
Yes, but we don't care about feasibility of scheduling at this layer, only applying high level backpressure.